test_util/helpers.py

```diff
@@ -23,81 +24,137 @@ def path_join(p1, p2):
+                must return the port for that string. Intended for communicating with an API
+                that uses different ports on different hosts (e.g. AdminRouter, Mesos )
         """
-        headers = copy.copy(self.default_headers)
+        self.url = url
+        self.session = requests.Session()
+        self._get_node_port = get_node_port
 
-        # allow kwarg to override verification so client can be used generically
-        if self.ca_cert_path and 'verify' not in kwargs:
-            kwargs['verify'] = self.ca_cert_path
+    def get_client(self, scheme=None, host=None, path=None, query=None, fragment=None, port=None):
+        """Takes the same arguments as Url.get_url
+        """
+        clone = copy.deepcopy(self)
+        clone.url = self.url.get_url(scheme=scheme, host=host, path=path, query=query, fragment=fragment, port=port)
+        return clone
 
-        if self.api_base:
-            path = path_join(self.api_base, path)
+    def api_request(self, method, path_ex, scheme=None, host=None, query=None,
+                    fragment=None, path=None, port=None, node=None, **kwargs):
+        """ Direct wrapper for session.request. Returns request.Response
+        Args:
+            method: the HTTP request method to be used
+            path_ex: the extension to the path that is set as the default Url
 
+        Keyword Args:
+            All the named keyword args except node are identical to Url.get_url
+            node: can only be used if a get_node_port is set.
+            **kwargs: anything that can be passed to requests.request
+        """
         if node is not None:
-            assert host_url is None, 'Cannot set both node ({}) and host_url ({})'.format(node, host_url)
-            assert self._get_node_url is not None, 'get_node_url function must be supplied'
-            host_url = self._get_node_url(node, port=port)
-        else:
-            host_url = host_url if host_url else self.default_host_url
-        request_url = path_join(host_url, path)
-        headers.update(kwargs.pop('headers', {}))
-        logging.info('Request method {}: {}'.format(method, request_url))
-        logging.debug('Reqeust kwargs: {}'.format(kwargs))
-        logging.debug('Request headers: {}'.format(headers))
-        return requests.request(method, request_url, headers=headers, **kwargs)
+            assert port is None, 'node is intended to retrieve port; cannot set both simultaneously'
+            assert host is None, 'node is intended to retrieve host; cannot set both simultaneously'
+            assert self._get_node_port is not None, 'get_node_netloc must be specified for this ApiClient!'
+            port = self._get_node_port(node)
+            # do not explicitly declare default ports
+            host = node
+            if (port == 80 and self.url.scheme == 'http') or (port == 443 and self.url.scheme == 'https'):
+                port = None
```

Comment:
As written there will be no way to use ApiClient with node keyword set and port=80, however when not using the node keyword, you can set port=80 or 443 all you desire.

For this reason of a somewhat awkard API, I originally implemented this with a get_node_url function instead of get_node_port. Given that all it was really adding was the port.

I think I should go back to get_node_url as to keep ApiClient as generic as possible